[#170824510] improvements tab deadlines

Edigio opened the pull request:
2 improvements:

1) remove all fake items between past deadlines

2) change the chronological order of the deadlines

digitalcitizenship commented:
<!--
  0 failure: 
  0 warning: 
  
  1 markdown notices
  DangerID: danger-id-default;
-->




## Affected stories

  * 🌟 [#170824510](https://www.pivotaltracker.com/story/show/170824510): Improvements Tab scadenze nella sezione messaggi

<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

Edigio commented:
Improvements:
- Removed all fake items
- Change the chronological order of the deadlines
- Added refresh control

<table style="width:100%">
  <tr>
    <th>Before</th>
    <th>After</th>
  </tr>
  <tr>
    <td><img src="https://user-images.githubusercontent.com/7993551/75875972-12be3180-5e15-11ea-90e5-97618a42d348.gif"/></td>
<td><img src="https://user-images.githubusercontent.com/7993551/75876050-3aad9500-5e15-11ea-93f4-846c0135f567.gif"/></td></tr>
</table>

valentinacala commented on ts/components/messages/MessageAgenda.tsx:
```diff
@@ -217,11 +210,9 @@ const generateItemLayouts = (sections: Sections) => {
       const isLastItem = dataIndex === data.length - 1;
 
-      const itemHeight = isFake ? FAKE_ITEM_HEIGHT : ITEM_HEIGHT;
+      const itemHeight = ITEM_HEIGHT;
```
you could avoid creating a new constant `itemHeight`

valentinacala commented on ts/components/messages/MessageAgenda.tsx:
```diff
@@ -252,82 +243,148 @@ const generateItemLayouts = (sections: Sections) => {
+        <Placeholder.Line
+          textSize={customVariables.fontSizeBase}
+          color={customVariables.shineColor}
+          width="75%"
```
```suggestion
          width={"75%"}
```

valentinacala commented on ts/components/messages/MessageAgenda.tsx:
```diff
@@ -252,82 +243,148 @@ const generateItemLayouts = (sections: Sections) => {
+          textSize={customVariables.fontSizeBase}
+          color={customVariables.shineColor}
+          width="75%"
+          animate="shine"
```
```suggestion
          animate={"shine"}
```

valentinacala commented on ts/components/messages/MessageAgenda.tsx:
```diff
@@ -252,82 +243,148 @@ const generateItemLayouts = (sections: Sections) => {
+          width="100%"
+          firstLineWidth="100%"
+          lastLineWidth="55%"
+          animate="shine"
```
use {} for all attribute values

valentinacala commented on ts/components/messages/MessageAgenda.tsx:
```diff
@@ -252,82 +243,148 @@ const generateItemLayouts = (sections: Sections) => {
+const SectionHeaderPlaceholder = (
+  <View style={styles.sectionHeaderWrapper}>
+    <View style={styles.sectionHeaderContent}>
+      <Placeholder.Line
```
it seems similar to [this](https://github.com/pagopa/io-app/pull/1561/files#diff-ea9cc6ac235ffba1edf4270a064eed7aR266). Is it make sense for you to create a common item to avoid code duplication?

valentinacala commented on ts/components/messages/MessageAgenda.tsx:
```diff
@@ -252,82 +243,148 @@ const generateItemLayouts = (sections: Sections) => {
+      this.state.isDeadlinesLoaded &&
+      this.sectionListRef.current !== null &&
+      this.state.isFirstLoading &&
+      this.props.sections !== undefined &&
```
Maybe the condition verifies that the item is not undefined if you write just:
```suggestion
      this.sectionListRef.current &&
      this.state.isFirstLoading &&
      this.props.sections &&
```
I'm less sure about null object, can you check it?

valentinacala commented on ts/components/messages/MessagesDeadlines.tsx:
```diff
@@ -581,41 +308,63 @@ class MessagesDeadlines extends React.PureComponent<Props, State> {
   }
 
+  /**
+   * Return number pof sections to load
```
```suggestion
   * Returns the number of sections to load
```

Undermaken commented on ts/components/messages/MessageAgenda.tsx:
```diff
@@ -36,6 +36,8 @@ const SECTION_HEADER_HEIGHT = 48;
 const FAKE_ITEM_HEIGHT = 75;
 const ITEM_SEPARATOR_HEIGHT = 1;
+const ITEM_WITHOUT_CTABAR_HEIGHT = 114;
+const ITEM_LOADING_HEIGHT = ITEM_WITHOUT_CTABAR_HEIGHT;
```
maybe you can use only a constant to represent the value 114

Undermaken commented on ts/components/messages/MessageAgenda.tsx:
```diff
@@ -344,45 +400,70 @@ class MessageAgenda extends React.PureComponent<Props, State> {
+      // tslint:disable-next-line: no-object-mutation
+      this.messageToLoadFromSections += messageInside;
+    });
+    this.setState({
```
```suggestion
    const numMessagesToRender =
      this.messageToLoadFromSections + this.state.numMessagesToRender;
    this.setState({
      numMessagesToRender
    });
```

Undermaken commented on ts/components/messages/MessageAgenda.tsx:
```diff
@@ -344,45 +400,70 @@ class MessageAgenda extends React.PureComponent<Props, State> {
+  private getMessageToLoadFromSections = (sections: Sections) => {
+    // tslint:disable-next-line: no-object-mutation
+    this.messageToLoadFromSections = 0;
+    sections.forEach(s => {
```
```typescript
private getMessageToLoadFromSections = (sections: Sections) => {
    const messageToLoadFromSections = sections.reduce(
      (acc, curr) => acc + curr.data.length,
      0
    );
    const numMessagesToRender =
      messageToLoadFromSections + this.state.numMessagesToRender;
    this.setState({
      numMessagesToRender
    });
    return messageToLoadFromSections;
  };
```

You can avoid using `this.messageToLoadFromSections`. 
Anyway you are returning a value from a class method which is held in a class member

Undermaken commented on ts/components/messages/MessagesDeadlines.tsx:
```diff
@@ -581,41 +308,63 @@ class MessagesDeadlines extends React.PureComponent<Props, State> {
+  /**
+   * Returns the number of sections to load
+   */
+  private sectionToLoad = (
```
```typescript
 /**
   * Returns the number of sections to load
   */
  private sectionToLoad = (
    messagesState: pot.Pot<ReadonlyArray<MessageState>, string>
  ): number => {
    return pot.getOrElse(
      pot.map(messagesState, value => {
        // check loading messages
        const isSomeLoading = value.find(m => pot.isLoading(m.message));
        if (isSomeLoading !== undefined) {
          return 0;
        }
        return value.filter(m => {
          const message = m.message;
          return (
            pot.isSome(message) && message.value.content.due_date !== undefined
          );
        }).length;
      }),
      0
    );
  };
```

Undermaken commented on ts/components/messages/MessageAgenda.tsx:
```diff
@@ -252,82 +242,148 @@ const generateItemLayouts = (sections: Sections) => {
-  public componentDidUpdate(prevProps: Props) {
-    // Change status loading to show progress
+  public async componentDidUpdate() {
     if (
```
```suggestion
    if (
      this.sectionListRef.current &&
      this.state.isFirstLoading &&
      this.state.isDeadlinesLoaded &&
      this.props.sections.length > 0
    )
```
please leave some comments to explain what is happening

Edigio commented:
@Undermaken I removed the timeout on the sectionlist scrollTo
